python/ray/data/_internal/block_batching.py

```diff
@@ -142,12 +154,56 @@ def batch_blocks(
         stats=stats,
     )
 
+    if prefetch_batches > 0:
+        batch_iter = _make_async_gen(batch_iter, prefetch_buffer_size=prefetch_batches)
+
     for formatted_batch in batch_iter:
         user_timer = stats.iter_user_s.timer() if stats else nullcontext()
         with user_timer:
             yield formatted_batch
 
 
+def _make_async_gen(
+    base_iterator: Iterator[T], prefetch_buffer_size: int = 1
+) -> Iterator[T]:
+    """Returns a new iterator with elements fetched from the base_iterator
+    in an sync fashion using a background thread.
+
+    Args:
+        base_iterator: The iterator to asynchronously fetch from.
+        prefetch_buffer_size: The maximum number of items to prefetch. Increasing the
+            size allows for more computation overlap for very expensive downstream udfs.
+            However it comes at the cost of additional memory overhead.
+
+    Returns:
+        An iterator with the same elements as the base_iterator.
+    """
+
+    # TODO: Add other async implementations: ThreadPool, Ray Actors.
+
+    fetch_queue = queue.Queue(maxsize=prefetch_buffer_size)
+
+    def _async_fetch():
+        for item in base_iterator:
+            fetch_queue.put(item, block=True)
```

Comment:
shall we throw exception if `item` is None here? Just to be a little bit defensive here for future-proof.

Reply:
great, thanks!